gnomad_qc/v4/sample_qc/relatedness.py

```diff
@@ -0,0 +1,251 @@
+            sample_list = mt_v3.s.collect() + mt_v4.s.collect()
+            if len(sample_list) != len(set(sample_list)):
+                raise DataException("duplicate samples found in v3 + v4")
+
+            # Check that the number of sites match.
+            row_count_v3 = mt_v3.count_rows()
+            row_count_v4 = mt_v4.count_rows()
+            if row_count_v3 != row_count_v4:
+                raise DataException("number of sites doesn't match between v3 and v4")
+
+            # Combine v3 and v4 dense QC matrices, as we need to compute relatedness
+            # across the union of samples.
+            mt = mt_v3.union_cols(mt_v4)
+
+            # Compute the field required for KING.
+            mt = mt.select_entries(n_alt_alleles=mt.GT.n_alt_alleles())
+
+            # Create a table based on entries alone. By adding a row and column index,
+            # we can drop the locus, alleles, and sample field, as well as skip writing
+            # missing entries.
+            mt = mt.select_globals()
+            mt = mt.add_row_index()
+            mt = mt.add_col_index()
+            entries = mt.entries()
+            entries = entries.key_by()
+            entries = entries.select(
+                entries.row_idx, entries.col_idx, entries.n_alt_alleles
+            )
+
+            # Export one compressed Parquet file per partition.
+            logger.info("Writing Parquet tables...")
+            entries.to_spark().write.option("compression", "zstd").parquet(input_path)
+
+            # Write metadata that's useful for postprocessing. Map `col_idx` to `s`
+            # explicitly so we don't need to rely on a particular order returned by
+            # `collect`.
+            logger.info("Writing metadata...")
+            col_idx_mapping = hl.struct(col_idx=mt.col_idx, s=mt.s).collect()
+            col_idx_mapping.sort(key=lambda e: e.col_idx)
+            metadata = {
+                "num_sites": row_count_v4,
+                "samples": [e.s for e in col_idx_mapping],
+            }
+            with hadoop_open(f"{input_path}/metadata.json", 'w') as f:
+                json.dump(metadata, f)
+
+        if args.create_relatedness_table:
+            from hail.utils.java import Env
+
+            spark = Env.spark_session()
+            df = spark.read.parquet(cuking_output_path(test=test))
+            ht = hl.Table.from_spark(df)
+            ht = ht.repartition(64)
+            ht = ht.key_by(ht.i, ht.j)
+            ht.write(relatedness.path)
+
+        if args.compute_related_samples_to_drop:
+            # compute_related_samples_to_drop uses a rank table as a tie breaker when
+            # pruning samples. We determine rankings as follows: filtered samples are
+            # ranked lowest, followed by non-releasable samples, followed by presence in
+            # v3 (favoring v3 over v4), and then by coverage.
+            def annotate_project_meta_ht(
```

Comment:
v4 will not have an `exclude`, and I don't think we need to worry about including it for v3 either since I think we mostly care about just making sure the `release` samples are prioritized. We can still keep releasable, we have that for both v3 and v4 so doesn't hurt to keep it

Reply:
A lot of the complexity here comes from merging the v3 and v4 project meta tables. If `qc.mt()` will contain both v3 + v4 samples, would it make sense to also create meta tables across both v3 and v4? I'd imagine that this would simplify any downstream analysis (including this code) quite a bit.

Reply:
I added a PR for this: https://github.com/broadinstitute/gnomad_qc/pulls

Reply:
Awesome, thanks very much! That simplifies all of this greatly.